File: src/credential-names.mjs
/**
 * The credential-noun vocabulary, and the NAME matcher built from it.
 *
 * `agent-sanitizer/credential-names` publishes the vocabulary as data so a
 * consumer derives its matcher from one list instead of forking one. That shares
 * the words but not the RULE, and the rule is where the mistakes are: whether a
 * noun must be the name's trailing segment or may sit anywhere in it, whether the
 * comparison folds case, whether a multi-word noun is compared as one run, and
 * whether the walk stays linear in the name's length. A consumer that renders the
 * vocabulary into one alternation regex gets a pattern with polynomial
 * backtracking on a long name; one that anchors on the trailing segment alone
 * matches nothing at all for `DEPLOY_TOKEN_ORG` or `OAUTH_TOKEN_FALLBACK_4`.
 *
 * So the mechanics live here and the POLICY stays the caller's, selected by
 * `scope`. The two scopes are not interchangeable and neither is a better
 * default:
 *
 * * `"trailing"` — the noun is the name's last underscore-delimited run. What a
 *   REDACTOR wants: it decides what to cut out of text a human will read, where
 *   over-matching mangles legitimate output.
 * * `"any-segment"` — the noun is any whole run of the name's segments. What an
 *   env-var SCRUB wants: it decides what a subprocess may inherit, where the two
 *   error directions are not symmetric — an unstripped credential leaks
 *   silently, an over-stripped variable breaks the command loudly.
 *
 * Matching is set membership over the name's underscore-delimited runs, never an
 * alternation of the nouns: 28 prefix-sharing renderings (`API_KEY`, `APIKEY`,
 * `ACCESS_KEY`, …) compile to a pattern a redos analyzer measures as polynomial,
 * and a matcher a hostile variable NAME can stall has a denial-of-service in
 * front of it. The run length is bounded by the longest noun, so the walk is
 * linear in the name's segment count rather than quadratic.
 *
 * The vocabulary itself is `python/agent_sanitizer/secrets/data/credential-names.json`
 * — the same file the Python renderings read and the same file the
 * `agent-sanitizer/credential-names` subpath exports, so the ecosystems cannot
 * drift apart on the words either.
 */
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { fileURLToPath } from "node:url";

const DATA_FILE = join(
  dirname(fileURLToPath(import.meta.url)),
  "..",
  "python",
  "agent_sanitizer",
  "secrets",
  "data",
  "credential-names.json",
);

const FILE_LABEL = "credential-names.json";
const ENV_NAME_USE = "env-name";
const FIELD_VALUE_USE = "field-value";
const KNOWN_USES = new Set([ENV_NAME_USE, FIELD_VALUE_USE]);

// a-z0-9 only is what lets a part interpolate into a consumer's pattern
// unescaped. Anchored with ^…$ over a single character class, so a part carrying
// a newline is rejected here rather than accepted and then rejected by the Python
// validator applying the same rule to the same file.
const PART_RE = /^[a-z0-9]+$/;

/** @param {string[]} values @returns {string[]} `values` without duplicates, first-occurrence order kept. */
function dedupe(values) {
  return [...new Set(values)];
}

/** The whole-name forms of `parts`: underscore-joined and bare-joined, upper-cased.
 *
 * Both are emitted because both spellings occur in the wild (`API_KEY` and
 * `APIKEY`) and a matcher comparing underscore-delimited runs sees them as
 * different tokens. A single-part noun collapses to one form.
 * @param {string[]} parts @returns {string[]} */
function segmentForms(parts) {
  return dedupe([parts.join("_").toUpperCase(), parts.join("").toUpperCase()]);
}

/** `value` as a validated array of noun parts, or throw naming `field`.
 * @param {unknown} value @param {string} field @returns {string[]} */
function parts(value, field) {
  if (!Array.isArray(value) || value.length === 0)
    throw new Error(`${FILE_LABEL}: ${field} is empty or missing`);
  const bad = value.filter(
    (part) => typeof part !== "string" || !PART_RE.test(part),
  );
  if (bad.length)
    throw new Error(
      `${FILE_LABEL}: bad part(s) ${JSON.stringify(bad)} in ${field}`,
    );
  return value;
}

/** `value` as a validated non-empty subset of the known uses, or throw.
 *
 * An unknown use is a refusal, not a skip: silently ignoring it would drop the
 * noun from every rendering, which is how a credential noun becomes inert.
 * @param {unknown} value @param {string} field @returns {Set<string>} */
function uses(value, field) {
  if (!Array.isArray(value) || value.length === 0)
    throw new Error(`${FILE_LABEL}: ${field} is empty or missing`);
  const unknown = value.filter((use) => !KNOWN_USES.has(use)).sort();
  if (unknown.length)
    throw new Error(
      `${FILE_LABEL}: unknown use(s) ${JSON.stringify(unknown)} in ${field}`,
    );
  return new Set(value);
}

/** @typedef {{ segments: string[], fieldNamePatterns: string[], nonSecretSegments: string[] }} CredentialNames */

/** Validate `spec` and return its renderings — the JavaScript twin of
 * `agent_sanitizer.secrets.parse_credential_names`, over the same file.
 *
 * A malformed spec throws rather than degrading. An empty list would render an
 * alternation that matches nothing (every credential forwarded verbatim) and a
 * part carrying a regex metacharacter one that matches everything (all output
 * blanked), so neither may reach a consumer's matcher.
 * @param {Record<string, unknown>} spec @returns {CredentialNames} */
export function parseCredentialNames(spec) {
  const nouns = spec?.nouns;
  if (!Array.isArray(nouns) || nouns.length === 0)
    throw new Error(`${FILE_LABEL}: nouns is empty or missing`);
  /** @type {string[]} */
  const segments = [];
  /** @type {string[]} */
  const fieldNamePatterns = [];
  nouns.forEach((noun, index) => {
    if (typeof noun !== "object" || noun === null || Array.isArray(noun))
      throw new Error(`${FILE_LABEL}: nouns[${index}] is not an object`);
    const nounParts = parts(noun.parts, `nouns[${index}].parts`);
    const nounUses = uses(noun.uses, `nouns[${index}].uses`);
    if (nounUses.has(ENV_NAME_USE)) segments.push(...segmentForms(nounParts));
    if (nounUses.has(FIELD_VALUE_USE))
      fieldNamePatterns.push(nounParts.join("[_-]?"));
  });

  const suffixes = spec?.nonSecretSuffixes;
  if (!Array.isArray(suffixes) || suffixes.length === 0)
    throw new Error(`${FILE_LABEL}: nonSecretSuffixes is empty or missing`);
  const nonSecretSegments = suffixes.flatMap((suffix, index) =>
    segmentForms(parts(suffix, `nonSecretSuffixes[${index}]`)),
  );

  // A vocabulary that renders nothing for one matcher would hand that consumer an
  // empty set, which matches nothing and forwards every credential.
  if (!segments.length)
    throw new Error(`${FILE_LABEL}: no noun is marked ${ENV_NAME_USE}`);
  if (!fieldNamePatterns.length)
    throw new Error(`${FILE_LABEL}: no noun is marked ${FIELD_VALUE_USE}`);
  return {
    segments: dedupe(segments),
    fieldNamePatterns: dedupe(fieldNamePatterns),
    nonSecretSegments: dedupe(nonSecretSegments),
  };
}

/** @type {CredentialNames | undefined} */
let _packaged;
/** The validated renderings of the packaged vocabulary, memoized.
 *
 * Read lazily, never at module load: a static importer that crashed at LOAD would
 * abort before its own fail-closed catch installs, and a guardrail that fails to
 * load is a guardrail that fails OPEN. Deferring to first use routes a missing or
 * corrupt data file into the caller's catch instead.
 * @returns {CredentialNames} */
export function credentialNames() {
  return (_packaged ??= parseCredentialNames(
    JSON.parse(readFileSync(DATA_FILE, "utf8")),
  ));
}

/** @typedef {"trailing" | "any-segment"} CredentialNameScope */

/** A predicate: does this env-var NAME hold a credential?
 *
 * `scope` selects the rule — `"trailing"` for a redactor (the noun must be the
 * name's last run), `"any-segment"` for an env scrub (the noun may be any run).
 * See this module's header for why that choice belongs to the caller.
 *
 * `declineNonSecret` applies the vocabulary's `nonSecretSuffixes`: a name ending
 * in one holds an identifier or a public key (`AWS_ACCESS_KEY_ID`), not a secret.
 * Leave it on for a redactor, where redacting an identifier out of output is a
 * visible defect; turn it off for a scrub whose failure to strip is the worse
 * error. It is applied to the name's trailing run under both scopes, because a
 * non-secret marker only means anything at the end of a name.
 *
 * The returned predicate closes over the parsed vocabulary, so build it once and
 * reuse it — the parse and validation are not repeated per name.
 *
 * @param {{ scope?: CredentialNameScope, declineNonSecret?: boolean, spec?: Record<string, unknown> }} [options]
 * @returns {(name: string) => boolean} */
export function credentialNameMatcher(options = {}) {
  const { scope = "trailing", declineNonSecret = true, spec } = options;
  if (scope !== "trailing" && scope !== "any-segment")
    throw new Error(
      `credentialNameMatcher: unknown scope ${JSON.stringify(scope)}`,
    );
  const vocabulary = spec ? parseCredentialNames(spec) : credentialNames();
  const nouns = new Set(vocabulary.segments);
  const nonSecret = new Set(vocabulary.nonSecretSegments);
  // The longest noun's run length. A run longer than this can match no noun, so
  // bounding the walk here is what keeps it linear in the name's segment count —
  // without it a 2000-underscore name costs quadratic time.
  const maxRun = Math.max(
    ...vocabulary.segments.map((noun) => noun.split("_").length),
  );
  /** @param {string[]} words @returns {string[]} */
  const trailingRuns = (words) =>
    Array.from({ length: Math.min(maxRun, words.length) }, (_, i) =>
      words.slice(words.length - (i + 1)).join("_"),
    );
  return (name) => {
    const words = name.toUpperCase().split("_");
    if (
      declineNonSecret &&
      trailingRuns(words).some((run) => nonSecret.has(run))
    )
      return false;
    if (scope === "trailing")
      return trailingRuns(words).some((run) => nouns.has(run));
    for (let start = 0; start < words.length; start++)
      for (let span = 1; span <= maxRun && start + span <= words.length; span++)
        if (nouns.has(words.slice(start, start + span).join("_"))) return true;
    return false;
  };
}

File: python/agent_sanitizer/secrets/credential_names.py
"""The published credential-noun vocabulary: the words that make a name a secret.

``data/credential-names.json`` is the single source; this module validates it and
renders it for the two matchers that need it, which are deliberately NOT unified:

* an env-var **NAME** matcher inspects no value, so it wants whole-name segments
  (``API_KEY``, ``APIKEY``) — :func:`credential_name_segments`;
* a ``field = value`` **redaction** matcher wants regex fragments tolerant of the
  separator the author wrote (``api[_-]?key``) — :func:`credential_field_name_patterns`,
  which the engine's ``_FIELD_NAMES`` alternation is built from.

Rendering the same nouns two ways is the point: unifying the *predicates* would
either weaken the name scrub or flood the redactor with false positives, so only
the vocabulary is shared. A noun the JSON marks ``env-name`` only is excluded from
the field-value rendering for exactly that reason.

A malformed spec raises rather than degrading: an empty list yields an alternation
that matches nothing (every credential forwarded verbatim) and a part carrying a
regex metacharacter yields one that matches everything (all output blanked), so
neither may reach a consumer's pattern builder.

The same file is published to JavaScript consumers as the npm subpath export
``agent-sanitizer/credential-names``.
"""

import functools
import json
import re
from collections.abc import Mapping, Sequence
from pathlib import Path
from typing import NamedTuple

CREDENTIAL_NAMES_FILE = (
    Path(__file__).resolve().parent / "data" / "credential-names.json"
)

# a-z0-9 only is what lets a part interpolate into a consumer's pattern unescaped.
# Matched with fullmatch, not `$`, because Python's `$` also matches just before a
# trailing newline — `"token\n"` must be rejected here, not accepted and then
# rejected by a JavaScript consumer applying the same rule to the same file.
_PART_RE = re.compile(r"[a-z0-9]+")

ENV_NAME_USE = "env-name"
FIELD_VALUE_USE = "field-value"
_USES = frozenset({ENV_NAME_USE, FIELD_VALUE_USE})

_FILE_LABEL = "credential-names.json"


class CredentialNames(NamedTuple):
    """The renderings of a validated vocabulary spec."""

    segments: tuple[str, ...]
    field_name_patterns: tuple[str, ...]
    non_secret_segments: tuple[str, ...]


def _dedupe(values: Sequence[str]) -> tuple[str, ...]:
    """``values`` with duplicates dropped, first-occurrence order preserved."""
    return tuple(dict.fromkeys(values))


def _segment_forms(parts: Sequence[str]) -> tuple[str, ...]:
    """The whole-name forms of ``parts``: underscore-joined and bare-joined, upper.

    Both are emitted because both spellings occur in the wild (``API_KEY`` and
    ``APIKEY``) and a name matcher anchored on underscore-delimited segments sees
    them as different tokens. A single-part noun collapses to one form.
    """
    return _dedupe(["_".join(parts).upper(), "".join(parts).upper()])


def _parts(value: object, field: str) -> tuple[str, ...]:
    """``value`` as a validated tuple of noun parts, or raise naming ``field``."""
    parts = tuple(value) if isinstance(value, (list, tuple)) else ()
    if not parts:
        raise ValueError(f"{_FILE_LABEL}: {field} is empty or missing")
    bad = [p for p in parts if not (isinstance(p, str) and _PART_RE.fullmatch(p))]
    if bad:
        raise ValueError(f"{_FILE_LABEL}: bad part(s) {bad} in {field}")
    return parts


def _uses(value: object, field: str) -> frozenset[str]:
    """``value`` as a validated non-empty subset of the known uses, or raise.

    An unknown use is a refusal, not a skip: silently ignoring it would drop the
    noun from every rendering, which is how a credential noun becomes inert.
    """
    uses = tuple(value) if isinstance(value, (list, tuple)) else ()
    if not uses:
        raise ValueError(f"{_FILE_LABEL}: {field} is empty or missing")
    unknown = sorted(set(uses) - _USES)
    if unknown:
        raise ValueError(f"{_FILE_LABEL}: unknown use(s) {unknown} in {field}")
    return frozenset(uses)


def _sequence(value: object, field: str) -> tuple[object, ...]:
    """``value`` as a validated non-empty tuple, or raise naming ``field``."""
    items = tuple(value) if isinstance(value, (list, tuple)) else ()
    if not items:
        raise ValueError(f"{_FILE_LABEL}: {field} is empty or missing")
    return items


def parse_credential_names(spec: Mapping[str, object]) -> CredentialNames:
    """Validate ``spec`` and return its renderings.

    Pure and public so a consumer — and the fail-closed tests — can drive the
    validator with a substitute spec instead of only through the packaged file.
    """
    segments: list[str] = []
    patterns: list[str] = []
    for index, noun in enumerate(_sequence(spec.get("nouns"), "nouns")):
        if not isinstance(noun, Mapping):
            raise ValueError(f"{_FILE_LABEL}: nouns[{index}] is not an object")
        parts = _parts(noun.get("parts"), f"nouns[{index}].parts")
        uses = _uses(noun.get("uses"), f"nouns[{index}].uses")
        if ENV_NAME_USE in uses:
            segments.extend(_segment_forms(parts))
        if FIELD_VALUE_USE in uses:
            patterns.append("[_-]?".join(parts))

    non_secret: list[str] = []
    for index, suffix in enumerate(
        _sequence(spec.get("nonSecretSuffixes"), "nonSecretSuffixes")
    ):
        non_secret.extend(_segment_forms(_parts(suffix, f"nonSecretSuffixes[{index}]")))

    # A vocabulary that renders nothing for one matcher would hand that consumer an
    # empty alternation, which matches nothing and forwards every credential.
    if not segments:
        raise ValueError(f"{_FILE_LABEL}: no noun is marked {ENV_NAME_USE}")
    if not patterns:
        raise ValueError(f"{_FILE_LABEL}: no noun is marked {FIELD_VALUE_USE}")
    return CredentialNames(
        segments=_dedupe(segments),
        field_name_patterns=_dedupe(patterns),
        non_secret_segments=_dedupe(non_secret),
    )


@functools.cache
def _rendered() -> CredentialNames:
    """The validated renderings of the packaged vocabulary. Raises if the data file
    is absent (fail closed — a partial vocabulary silently under-matches)."""
    return parse_credential_names(
        json.loads(CREDENTIAL_NAMES_FILE.read_text(encoding="utf-8"))
    )


def credential_name_segments() -> tuple[str, ...]:
    """Upper-case whole-name forms of every noun usable against a variable NAME.

    A name whose trailing underscore-delimited segment is one of these holds a
    credential (``DEPLOY_API_KEY``), so a consumer matching by trailing segment
    self-populates from this list instead of curating its own.
    """
    return _rendered().segments


def credential_field_name_patterns() -> tuple[str, ...]:
    """Regex fragments for every noun usable as a ``field = value`` field name.

    Each fragment tolerates the separator the author wrote (``api_key``,
    ``api-key``, ``apikey``) and carries no metacharacter of its own, so a caller
    may join them into an alternation directly.
    """
    return _rendered().field_name_patterns


def non_secret_name_segments() -> tuple[str, ...]:
    """Upper-case whole-name forms of the trailing words that mark a
    credential-shaped name as holding a NON-secret (``KEY_ID``, ``PUBLIC_KEY``).

    A consumer matching by trailing segment must decline these, or it redacts an
    identifier or a public key out of legitimate output.
    """
    return _rendered().non_secret_segments
